### PR TITLE
InMemoryEventBus Enhancement #147 - store integration event handlers in dictionary instead of list

### PR DESCRIPTION
**Pull request by kgrzybek**

### CODE REVIEW

File: src/BuildingBlocks/EventBus/InMemoryEventBus.cs [hunk 1]
```diff
@@ -13,46 +12,50 @@ static InMemoryEventBus()
...
+            }
         }
 
         public async Task Publish<T>(T @event)
             where T : IntegrationEvent
         {
-            var eventType = @event.GetType();
+            var eventType = @event.GetType().FullName;
+
+            if (eventType == null)
```
Comment (AndreiGanichev) [inline src/BuildingBlocks/EventBus/InMemoryEventBus.cs:45]: when the condition is true? I think only when the argument is null so we should add null check to `@event?.GetType().FullName`.

File: src/BuildingBlocks/EventBus/InMemoryEventBus.cs [hunk 2]
```diff
@@ -13,46 +12,50 @@ static InMemoryEventBus()
...
 
-        private readonly List<HandlerSubscription> _handlers;
+        private readonly IDictionary<string, List<IIntegrationEventHandler>> _handlersDictionary;
 
         public void Subscribe<T>(IIntegrationEventHandler<T> handler)
             where T : IntegrationEvent
         {
-            _handlers.Add(new HandlerSubscription(handler, typeof(T).FullName));
+            var eventType = typeof(T).FullName;
+            if (eventType != null)
```
Comment (AndreiGanichev) [inline src/BuildingBlocks/EventBus/InMemoryEventBus.cs:26]: It's true always, isn't it?

**Review by AndreiGanichev — COMMENTED** [review]
`Publish` and `Subscribe` methods became more complicated imho. But we got rid of `HandlerSubscription` class and in general `Dictionary` looks fine.

File: src/BuildingBlocks/EventBus/InMemoryEventBus.cs (hunk 1 above)
Comment (kgrzybek) [reply src/BuildingBlocks/EventBus/InMemoryEventBus.cs:45]: `FullName` can be `null` in some cases: https://docs.microsoft.com/en-us/dotnet/api/system.type.fullname?view=net-5.0#System_Type_FullName

Of course in our case shouldn't be `null`, but still, it is better to add this null check imo.

File: src/BuildingBlocks/EventBus/InMemoryEventBus.cs (hunk 2 above)
Comment (kgrzybek) [reply src/BuildingBlocks/EventBus/InMemoryEventBus.cs:26]: The same as above - can be null.

**Pull request merged**